core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt

```diff
@@ -1,33 +1,32 @@
 package com.unciv.ui.pickerscreens
 
+import com.badlogic.gdx.scenes.scene2d.ui.Table
 import com.badlogic.gdx.scenes.scene2d.ui.TextField
 import com.unciv.logic.map.MapUnit
-import com.unciv.ui.utils.CameraStageBaseScreen
-import com.unciv.ui.utils.Popup
-import com.unciv.ui.utils.onClick
-import com.unciv.ui.utils.toTextButton
+import com.unciv.ui.utils.*
 
 class RenameUnitPopup(unit: MapUnit, screen: CameraStageBaseScreen): Popup(screen) {
 
     init {
-        addGoodSizedLabel("Choose name for [${unit.baseUnit.name}]").row()
+        val wrapper = Table()
+        wrapper.add(ImageGetter.getUnitIcon(unit.name).surroundWithCircle(80f)).padRight(10f)
+        wrapper.add("Choose name for [${unit.baseUnit.name}]".toLabel())
+        add(wrapper).colspan(2).row()
 
         val nameField = TextField("", skin)
-        // Disallowed special characters determined with US keyboard and a certain binary
-        // copyrighted by Firaxis Games
-        val illegal_chars = "%*[]\"\\|<>/?"
-        nameField.textFieldFilter = TextField.TextFieldFilter { _, char -> !(char in illegal_chars)}
-        // Max name length decided arbitrarily
-        nameField.maxLength = 10
-        add(nameField)
-        row()
-        add("OK".toTextButton().onClick {
-            if (nameField.text != "") {
+        // Disallowed special characters - used by tr() or simply precaution
+        val illegalChars = "[]{}\"\\<>"
+        nameField.textFieldFilter = TextField.TextFieldFilter { _, char -> char !in illegalChars}
+        nameField.maxLength = unit.civInfo.gameInfo.ruleSet.units.values.maxOf { it.name.length }
+        add(nameField).growX().colspan(2).row()
+        addOKButton {
+            if (nameField.text != "" && nameField.text != unit.name) {
```

Comment:
So you can't name a unit back to it's original name if you have already changed it once?